torchdata/dataloader2/dataloader2.py

```diff
@@ -130,6 +133,9 @@ def __iter__(self) -> Iterator[T_co]:
         self.valid_iterator_id = 0 if self.valid_iterator_id is None else self.valid_iterator_id + 1
         return DataLoader2Iterator(self, self.valid_iterator_id)
 
+    def __len__(self):
+        return len(self.datapipe)
+
```

Comment:
Have you tested how it returns when sharding_filter presents? At least in MP environment.

Reply:
No I have not but I can do that. See my other comment about potentially delegating to ReadingService for a similar reason.

Reply:
`sharding_filter` has not impact on the length, because length doesn't change until `apply_sharding` and I think that is done within processes and not the main one? Correct me if I am wrong on that.